### ISSUE TITLE
QSAR_Modelling_using_Streamlit.py

### ISSUE THREAD

**Issue by KOUSHIK-L** [PonnaiyasCAG/Build-CAG-QSAR#1]
import streamlit as st
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from rdkit import Chem
from rdkit.Chem import rdFingerprintGenerator
from sklearn.feature_selection import VarianceThreshold
from sklearn.ensemble import RandomForestRegressor
from sklearn.svm import SVR
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import mean_squared_error, r2_score
import xgboost as xgb

from tensorflow import keras
from tensorflow.keras.models import Model
from tensorflow.keras.callbacks import EarlyStopping, ModelCheckpoint
from tensorflow.keras import layers
import tensorflow as tf

import pickle
import torch
from torch_geometric.data import Data
from torch_geometric.nn import GCNConv, global_mean_pool
from torch_geometric.loader import DataLoader

from rdkit.Chem import rdmolops

from tqdm import tqdm

def mol_to_graph(smiles):
    mol = Chem.MolFromSmiles(smiles)
    if mol is None:
        return None

    num_atoms = mol.GetNumAtoms()
    features = []

    for atom in mol.GetAtoms():
        features.append([atom.GetAtomicNum()])

    edge_index = []
    for bond in mol.GetBonds():
        i = bond.GetBeginAtomIdx()
        j = bond.GetEndAtomIdx()
        edge_index.append([i, j])
        edge_index.append([j, i])  # bidirectional

    if len(edge_index) == 0:
        return None  # handle molecules with no bonds

    x = torch.tensor(features, dtype=torch.float)
    edge_index = torch.tensor(edge_index, dtype=torch.long).t().contiguous()

    return Data(x=x, edge_index=edge_index)

class GCN(torch.nn.Module):
    def __init__(self, hidden_channels):
        super(GCN, self).__init__()
        self.conv1 = GCNConv(1, hidden_channels)
        self.conv2 = GCNConv(hidden_channels, hidden_channels)
        self.linear = torch.nn.Linear(hidden_channels, 1)

    def forward(self, x, edge_index, batch):
        x = self.conv1(x, edge_index)
        x = torch.relu(x)
        x = self.conv2(x, edge_index)
        x = torch.relu(x)
        x = global_mean_pool(x, batch)
        return self.linear(x)

st.set_page_config(page_title="QSAR Regression Trainer", layout="centered")
st.title("QSAR Regression Model Trainer")

# uploaded_file = st.file_uploader("Upload your QSAR dataset (.csv or .tsv)", type=["csv", "tsv", "txt", ".xlsx"])

# if uploaded_file is not None:
#     try:
#         if uploaded_file.name.endswith((".tsv", ".csv", ".xlsx")):
#             df = pd.read_csv(uploaded_file, sep='\t')
#         else:
#             df = pd.read_csv(uploaded_file)

uploaded_file = st.file_uploader("Upload your QSAR dataset (.csv, .tsv, or .xlsx)", type=["csv", "tsv", "txt", "xlsx"])

if uploaded_file is not None:
    try:
        if uploaded_file.name.endswith(".xlsx"):
            df = pd.read_excel(uploaded_file)  # Reads .xlsx files
        elif uploaded_file.name.endswith((".tsv", ".csv")):
            df = pd.read_csv(uploaded_file, sep='\t')  # Reads .tsv and .csv files
        else:
            df = pd.read_csv(uploaded_file)  # Reads other .csv or .txt files

    # except Exception as e:
    #     st.error(f"An error occurred: {e}")

        st.success("File uploaded successfully!")
        st.write("Raw Dataset Preview:")
        st.dataframe(df.head())

        st.subheader("Data Preprocessing")
        missing_ratio = (df.isnull().sum() / len(df)) * 100
        st.write("Missing Values (%):")
        st.dataframe(missing_ratio[missing_ratio > 0.8])

        # def convert_to_M(row):
        #     value = row["Activity CC50 (uM)"]
        #     # unit = row["Activity CC50 (uM)"]
            
        #     try:
        #         if not None:
        #             return value / 1e9
        #         else:
        #             return np.nan
        #     except:
        #         return np.nan

        # df["Molar Concentration"] = df.apply(convert_to_M, axis=1)
        # df["Computed pIC50"] = -np.log10(df["Molar Concentration"])

        # if "pIC50" in df.columns:
        #     df["exp."] = df["pIC50"].combine_first(df["Computed pIC50"])
        # else:
        #     df["exp."] = df["Computed pIC50"]

        # st.success("pIC50 values computed successfully!")
 
        st.dataframe(df[["canonical smiles", "exp."]].head())

        st.subheader("Morgan Fingerprint Generation")
        morgan_gen = rdFingerprintGenerator.GetMorganGenerator(radius=2, fpSize=2048)

        def smiles_to_fingerprint(smiles):
            try:
                mol = Chem.MolFromSmiles(smiles)
                if mol is not None:
                    fingerprint = morgan_gen.GetFingerprint(mol)
                    return list(fingerprint)
                else:
                    return None
            except:
                return None

        df["Fingerprint"] = df["canonical smiles"].apply(smiles_to_fingerprint)
        df = df.dropna(subset=["Fingerprint", "exp."])

        st.success(f"Fingerprints generated! Valid compounds: {len(df)}")
        st.write("Fingerprint shape:", len(df["Fingerprint"].iloc[0]))

        X = np.array(df["Fingerprint"].to_list())
        y = df["exp."].values

        st.write("Features shape:", X.shape)
        st.write("Target shape:", y.shape)

        st.subheader("Feature Selection")
        selector = VarianceThreshold(threshold=0.01)
        X_selected = selector.fit_transform(X)

        correlation = np.corrcoef(X_selected.T, y)[-1, :-1]
        selected_features = np.abs(correlation) > 0.2
        X_filtered = X_selected[:, selected_features]

        rf = RandomForestRegressor(n_estimators=100, random_state=42)
        rf.fit(X_selected, y)
        important_indices = np.argsort(rf.feature_importances_)[-50:]
        X_important = X_selected[:, important_indices]

        st.success("Feature Selection Completed")

        st.subheader("Feature Selection Visualizations")
        fig, ax = plt.subplots()
        sns.heatmap(np.corrcoef(X_selected[:, important_indices].T), cmap="coolwarm", ax=ax)
        st.pyplot(fig)

        feat_imp = rf.feature_importances_[important_indices]
        fig2, ax2 = plt.subplots()
        ax2.bar(range(len(feat_imp)), feat_imp)
        ax2.set_title("Top 50 Feature Importances")
        st.pyplot(fig2)

        X_train, X_test, y_train, y_test = train_test_split(X_important, y, test_size=0.2, random_state=42)

        st.subheader("Model Training")
        model_choice = st.selectbox("Choose a Model", ["Random Forest", "SVM", "XGBoost", "Deep Neural Network", "Graph Neural Network"])

        if model_choice == "Random Forest":
            rf_model = RandomForestRegressor()
            params = {"n_estimators": [100, 200], "max_depth": [10, 20, None]}
            grid = GridSearchCV(rf_model, params, cv=3)
            grid.fit(X_train, y_train)
            model = grid.best_estimator_

        elif model_choice == "SVM":
            svm = SVR()
            params = {"C": [0.1, 1, 10], "gamma": ["scale", "auto"]}
            grid = GridSearchCV(svm, params, cv=3)
            grid.fit(X_train, y_train)
            model = grid.best_estimator_

        elif model_choice == "XGBoost":
            xgb_model = xgb.XGBRegressor()
            params = {"n_estimators": [100, 200], "max_depth": [3, 6]}
            grid = GridSearchCV(xgb_model, params, cv=3)
            grid.fit(X_train, y_train)
            model = grid.best_estimator_

        elif model_choice == "Deep Neural Network":
            model =  tf.keras.Sequential([
                layers.Dense(128, activation='relu', input_shape=(X_train.shape[1],)),
                layers.Dense(64, activation='relu'),
                layers.Dense(1)
            ])

            model.compile(
                optimizer=tf.keras.optimizers.Adam(learning_rate=1e-4),
                loss=tf.keras.losses.Huber(),
                metrics=["mse", "mae"]
            )

            checkpoint_filepath = 'egfr_regressor_model.keras'
            checkpointer = ModelCheckpoint(
                filepath=checkpoint_filepath,
                monitor="val_loss",
                mode="min",
                save_best_only=True,
                verbose=1
            )

            early_stopping = EarlyStopping(monitor="val_loss", patience=15, restore_best_weights=True)

            history = model.fit(
                X_train, y_train,
                epochs=100,
                batch_size=32,
                validation_data=(X_test, y_test),
                callbacks=[checkpointer, early_stopping]
            )

        elif model_choice == "Graph Neural Network":
            st.info("Training Graph Neural Network...")

            graph_list = []
            y_graph = []

            for i in range(len(df)):
                graph = mol_to_graph(df.iloc[i]["canonical smiles"])
                if graph is not None:
                    graph.y = torch.tensor([df.iloc[i]["exp."]], dtype=torch.float)
                    graph_list.append(graph)

            train_graphs, test_graphs = train_test_split(graph_list, test_size=0.2, random_state=42)
            train_loader = DataLoader(train_graphs, batch_size=32, shuffle=True)
            test_loader = DataLoader(test_graphs, batch_size=32, shuffle=False)

            device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
            model = GCN(hidden_channels=64).to(device)
            optimizer = torch.optim.Adam(model.parameters(), lr=0.001)
            loss_fn = torch.nn.MSELoss()

            for epoch in tqdm(range(100), desc="Epoch Progress"):  # Adding progress bar for epochs
                model.train()
                for batch in tqdm(train_loader, desc=f"Training Epoch {epoch+1}"):  # Progress within batches
                    batch = batch.to(device)
                    optimizer.zero_grad()
                    out = model(batch.x, batch.edge_index, batch.batch)
                    loss = loss_fn(out.view(-1), batch.y.view(-1))
                    loss.backward()
                    optimizer.step()

            # Evaluation
            model.eval()
            preds, truths = [], []
            with torch.no_grad():
                for batch in test_loader:
                    batch = batch.to(device)
                    out = model(batch.x, batch.edge_index, batch.batch)
                    preds.extend(out.view(-1).cpu().numpy())
                    truths.extend(batch.y.view(-1).cpu().numpy())

            rmse = mean_squared_error(truths, preds)
            r2 = r2_score(truths, preds)

            st.write(f"RMSE: {rmse:.4f}")
            st.write(f"R² Score: {r2:.4f}")

            fig3, ax3 = plt.subplots()
            ax3.scatter(truths, preds)
            ax3.set_xlabel("Actual pIC50")
            ax3.set_ylabel("Predicted pIC50")
            ax3.set_title("Actual vs Predicted (GNN)")
            st.pyplot(fig3)

            # Save model
            torch.save(model.state_dict(), "gnn_qsar_model.pth")
            # st.success("GNN model trained and saved as gnn_qsar_model.pth")

        if model_choice != "Graph Neural Network":
            y_pred = model.predict(X_test)
            rmse = mean_squared_error(y_test, y_pred)
            r2 = r2_score(y_test, y_pred)
            st.write(f"RMSE: {rmse:.4f}")
            st.write(f"R² Score: {r2:.4f}")

            fig3, ax3 = plt.subplots()
            ax3.scatter(y_test, y_pred)
            ax3.set_xlabel("Actual pIC50")
            ax3.set_ylabel("Predicted pIC50")
            ax3.set_title("Actual vs Predicted")
            st.pyplot(fig3)

        if model_choice == "Graph Neural Network":
            # torch.save(model.state_dict(), "gnn_qsar_model.pth")
            st.success("GNN model trained and saved as gnn_qsar_model.pth")
        else:	
            model_filename = f"{model_choice.replace(' ', '_').lower()}_qsar_model.pkl"
            pickle.dump(model, open(model_filename, 'wb'))
            st.success(f"{model_choice} model trained and saved as {model_filename}")

    except Exception as e:
        st.error(f"Error reading or processing file: {e}")
